core/silkworm/crypto/sha-256.c

```diff
@@ -23,99 +50,426 @@
+             * initial state.
+             */
+
+            /*
+             * create a 64-entry message schedule array w[0..63] of 32-bit words
+             * (The initial values in w[0..63] don't matter, so many implementations zero them here)
+             * copy chunk into first 16 words w[0..15] of the message schedule array
+             */
+            uint32_t w[16];
+
+            for (j = 0; j < 16; j++) {
+                if (i == 0) {
+                    w[j] = (uint32_t)p[0] << 24 | (uint32_t)p[1] << 16 | (uint32_t)p[2] << 8 | (uint32_t)p[3];
+                    p += 4;
+                } else {
+                    /* Extend the first 16 words into the remaining 48 words w[16..63] of the message schedule array: */
+                    const uint32_t s0 =
+                        right_rot(w[(j + 1) & 0xf], 7) ^ right_rot(w[(j + 1) & 0xf], 18) ^ (w[(j + 1) & 0xf] >> 3);
+                    const uint32_t s1 =
+                        right_rot(w[(j + 14) & 0xf], 17) ^ right_rot(w[(j + 14) & 0xf], 19) ^ (w[(j + 14) & 0xf] >> 10);
+                    w[j] = w[j] + s0 + w[(j + 9) & 0xf] + s1;
+                }
+                const uint32_t s1 = right_rot(ah[4], 6) ^ right_rot(ah[4], 11) ^ right_rot(ah[4], 25);
+                const uint32_t ch = (ah[4] & ah[5]) ^ (~ah[4] & ah[6]);
+                const uint32_t temp1 = ah[7] + s1 + ch + k[i << 4 | j] + w[j];
+                const uint32_t s0 = right_rot(ah[0], 2) ^ right_rot(ah[0], 13) ^ right_rot(ah[0], 22);
+                const uint32_t maj = (ah[0] & ah[1]) ^ (ah[0] & ah[2]) ^ (ah[1] & ah[2]);
+                const uint32_t temp2 = s0 + maj;
+
+                ah[7] = ah[6];
+                ah[6] = ah[5];
+                ah[5] = ah[4];
+                ah[4] = ah[3] + temp1;
+                ah[3] = ah[2];
+                ah[2] = ah[1];
+                ah[1] = ah[0];
+                ah[0] = temp1 + temp2;
+            }
+        }
+
+        /* Add the compressed chunk to the current hash value: */
+        for (i = 0; i < 8; i++) {
+            h[i] += ah[i];
+        }
+    }
 }
 
+static void sha_256_generic(uint32_t h[8], const void *input, size_t len) { sha_256_implementation(h, input, len); }
+
+static void (*sha_256_best)(uint32_t h[8], const void *input, size_t len) = sha_256_generic;
+
+#if defined(__x86_64__) && __has_attribute(target)
+
+__attribute__((target("bmi,bmi2"))) static void sha_256_x86_bmi(uint32_t h[8], const void *input, size_t len) {
+    sha_256_implementation(h, input, len);
+}
+
+// The following function was adapted from https://github.com/noloader/SHA-Intrinsics
+/*   Intel SHA extensions using C intrinsics               */
+/*   Written and place in public domain by Jeffrey Walton  */
+/*   Based on code from Intel, and by Sean Gulley for      */
+/*   the miTLS project.                                    */
+__attribute__((target("sha,sse4.1"))) static void sha_256_x86_sha(uint32_t h[8], const void *input, size_t len) {
```

Comment:
Did you try to use function multi-versioning? https://gcc.gnu.org/onlinedocs/gcc/Function-Multiversioning.html

Reply:
> Did you try to use function multi-versioning? https://gcc.gnu.org/onlinedocs/gcc/Function-Multiversioning.html

I've just tried and my clang say "error: function multiversioning doesn't support feature 'sha'"